src/internal/SetupResumeAcceptor.cpp

```diff
@@ -198,13 +206,42 @@ void SetupResumeAcceptor::closeAndRemoveConnection(
     const std::shared_ptr<FrameTransport>& transport,
     folly::exception_wrapper ex) {
   transport->close(ex);
-  connections_.erase(transport);
+  connections_->erase(transport);
 }
 
 void SetupResumeAcceptor::removeConnection(
     const std::shared_ptr<FrameTransport>& transport) {
   transport->setFrameProcessor(nullptr);
-  connections_.erase(transport);
+  connections_->erase(transport);
+}
+
+folly::Future<folly::Unit> SetupResumeAcceptor::close() {
+  folly::Promise<folly::Unit> closingPromise;
+  auto closingFuture = closingPromise.getFuture();
+
+  if(eventBase_->isInEventBaseThread()) {
+    closeAllConnections();
+    closingPromise.setValue();
+  } else {
+    auto queued = eventBase_->runInEventBaseThread([this, closingPromise = std::move(closingPromise)]() mutable {
+      closeAllConnections();
+      closingPromise.setValue();
+    });
+    CHECK(queued);
+  }
+  return closingFuture;
+}
+
+void SetupResumeAcceptor::closeAllConnections() {
+  if(!connections_) {
+    return;
+  }
+
+  folly::exception_wrapper closingEx = std::runtime_error("shutting down");
+  for(auto& connection : *connections_) {
+    connection->close(closingEx);
```

Comment:
I think it's better to just do `connection->close(std::runtime_error("shutting down"))`.  Otherwise you end up copying a lot of `folly::exception_wrapper` objects which I think is marginally more expensive than constructing them and passing them as rvalues.